routers/api/v1/api.go

```diff
@@ -189,84 +189,119 @@ func reqBasicAuth() macaron.Handler {
 	return func(ctx *context.Context) {
-		if !ctx.IsSigned || !ctx.User.IsAdmin {
+		if !isSiteAdmin(ctx) {
 			ctx.Error(403)
 			return
 		}
 	}
 }
 
-// reqOwner user should be the owner of the repo.
+// reqOwner user should be the owner of the repo or site admin.
 func reqOwner() macaron.Handler {
 	return func(ctx *context.Context) {
-		if !ctx.Repo.IsOwner() {
+		if !isRepoOwner(ctx) && !isSiteAdmin(ctx) {
 			ctx.Error(403)
 			return
 		}
 	}
 }
 
-// reqAdmin user should be an owner or a collaborator with admin write of a repository
+// reqAdmin user should be an owner or a collaborator with admin write of a repository, or site admin
 func reqAdmin() macaron.Handler {
 	return func(ctx *context.Context) {
-		if !ctx.Repo.IsAdmin() {
+		if !isRepoAdmin(ctx) && !isSiteAdmin(ctx) {
 			ctx.Error(403)
 			return
 		}
 	}
 }
 
-func reqRepoReader(unitType models.UnitType) macaron.Handler {
+// reqRepoWriter user should have a permission to write to a repo, or be a site admin
+func reqRepoWriter(unitTypes ...models.UnitType) macaron.Handler {
 	return func(ctx *context.Context) {
-		if !ctx.Repo.CanRead(unitType) {
+		if !isRepoWriter(ctx, unitTypes) && !isRepoAdmin(ctx) && !isSiteAdmin(ctx) {
 			ctx.Error(403)
 			return
 		}
 	}
 }
 
-func reqAnyRepoReader() macaron.Handler {
+// reqRepoReader user should have specific read permission or be a repo admin or a site admin
+func reqRepoReader(unitType models.UnitType) macaron.Handler {
 	return func(ctx *context.Context) {
-		if !ctx.Repo.HasAccess() {
+		if !isRepoReaderSpecific(ctx, unitType) && !isRepoAdmin(ctx) && !isSiteAdmin(ctx) {
 			ctx.Error(403)
 			return
 		}
 	}
 }
 
-func reqRepoWriter(unitTypes ...models.UnitType) macaron.Handler {
+// reqAnyRepoReader user should have any permission to read repository or permissions of site admin
+func reqAnyRepoReader() macaron.Handler {
 	return func(ctx *context.Context) {
-		for _, unitType := range unitTypes {
-			if ctx.Repo.CanWrite(unitType) {
-				return
-			}
+		if !isRepoReaderAny(ctx) && !isSiteAdmin(ctx) {
+			ctx.Error(403)
+			return
 		}
-
-		ctx.Error(403)
 	}
 }
 
-func reqOrgMembership() macaron.Handler {
+// reqOrgOwnership user should be an organization owner, or a site admin
+func reqOrgOwnership() macaron.Handler {
 	return func(ctx *context.APIContext) {
+		if isSiteAdmin(ctx.Context) {
+			return
+		}
+
 		var orgID int64
 		if ctx.Org.Organization != nil {
 			orgID = ctx.Org.Organization.ID
 		} else if ctx.Org.Team != nil {
 			orgID = ctx.Org.Team.OrgID
 		} else {
-			ctx.Error(500, "", "reqOrgMembership: unprepared context")
+			ctx.Error(500, "", "reqOrgOwnership: unprepared context")
 			return
 		}
 
-		if isMember, err := models.IsOrganizationMember(orgID, ctx.User.ID); err != nil {
-			ctx.Error(500, "IsOrganizationMember", err)
-			return
-		} else if !isMember {
+		isOwner, err := models.IsOrganizationOwner(orgID, ctx.User.ID)
+		if err != nil {
+			ctx.Error(500, "IsOrganizationOwner", err)
```

Comment:
There should be return after this

Reply:
Fixed!

May I ask, why it should be there when there is no other possible control flow? Is it a convention in Go or something more general? Thanks.

Reply:
Sry, I was looking only at diff and as return was before so I thought it could be some flow further. In anyway it is safer as if code gets added you don't need to check if all control flow branches did terminate correctly